NativeAOT: TypePreinit.ArrayInstance.TryStoreElement fails with invalid destination array size

Hi,

We're trying to move from our old CoreRT-based build setup to most recent NativeAOT bits.

I'm getting this error in the midst of native code generation.
What's the best way to troubleshoot where this error comes from / help provide info?

This is a 2500+ file project, so narrowing it down manually might take a lot of time.

```
Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray') [/Users/chr/d/sf/pkg/sfbackend/src/sfbackend.csproj]
  System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
     at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
     at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length)
     at ILCompiler.TypePreinit.ArrayInstance.TryStoreElement(Int32 index, Value value)
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL methodIL, Value[] parameters, Stack`1 recursionProtect, Int32& instructionCounter, Value& returnValue)
     at ILCompiler.TypePreinit.ScanType(CompilationModuleGroup compilationGroup, ILProvider ilProvider, MetadataType type)
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key)
     at ILCompiler.PreinitializationManager.IsPreinitialized(MetadataType type)
     at ILCompiler.DependencyAnalysis.NodeFactory.<CreateNodeCaches>b__39_3(MetadataType type)
     at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
     at ILCompiler.UsageBasedMetadataManager.GetRuntimeMappingDependenciesDueToReflectability(DependencyList& dependencies, NodeFactory factory, TypeDesc type)
     at ILCompiler.MetadataManager.GetDependenciesDueToReflectability(DependencyList& dependencies, NodeFactory factory, TypeDesc type)
     at ILCompiler.DependencyAnalysis.ConstructedEETypeNode.ComputeNonRelocationBasedDependencies(NodeFactory factory)
     at ILCompiler.DependencyAnalysis.ObjectNode.GetStaticDependencies(NodeFactory factory)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1 node)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack()
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes()
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan()
     at ILCompiler.Program.Run(String[] args)
     at ILCompiler.Program.Main(String[] args)
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/6.0.0-preview.6.21278.2/build/Microsoft.NETCore.Native.targets(282,5): error MSB3073: The command ""/Users/chr/.nuget/packages/runtime.osx-x64.microsoft.dotnet.ilcompiler/6.0.0-preview.6.21278.2/tools/ilc" @"obj/Cloud/net5.0/osx-x64/native/sfbackend.ilc.rsp"" exited with code 1. [/Users/chr/d/sf/pkg/sfbackend/src/sfbackend.csproj]
```

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs
@@ -1528,22 +1528,22 @@ namespace ILCompiler
                         Push(StackValueKind.Int32, ValueTypeValue.FromInt32(value.AsInt16())); break;
                     case TypeFlags.Int32:
                     case TypeFlags.UInt32:
-                        Push(StackValueKind.Int32, value); break;
+                        Push(StackValueKind.Int32, value.Clone()); break;
                     case TypeFlags.Int64:
                     case TypeFlags.UInt64:
-                        Push(StackValueKind.Int64, value); break;
+                        Push(StackValueKind.Int64, value.Clone()); break;
                     case TypeFlags.IntPtr:
                     case TypeFlags.UIntPtr:
                     case TypeFlags.Pointer:
                     case TypeFlags.FunctionPointer:
-                        Push(StackValueKind.NativeInt, value); break;
+                        Push(StackValueKind.NativeInt, value.Clone()); break;
                     case TypeFlags.Single:
                         Push(StackValueKind.Float, ValueTypeValue.FromDouble(value.AsSingle())); break;
                     case TypeFlags.Double:
-                        Push(StackValueKind.Float, value); break;
+                        Push(StackValueKind.Float, value.Clone()); break;
                     case TypeFlags.ValueType:
                     case TypeFlags.Nullable:
-                        Push(StackValueKind.ValueType, value); break;
+                        Push(StackValueKind.ValueType, value.Clone()); break;
                     case TypeFlags.Class:
                     case TypeFlags.Interface:
                     case TypeFlags.Array:
@@ -1759,6 +1759,7 @@ namespace ILCompiler
             public virtual long AsInt64() => ThrowInvalidProgram<long>();
             public virtual float AsSingle() => ThrowInvalidProgram<float>();
             public virtual double AsDouble() => ThrowInvalidProgram<double>();
+            public virtual Value Clone() => ThrowInvalidProgram<Value>();
         }
 
         private abstract class BaseValueTypeValue : Value
@@ -1782,6 +1783,11 @@ namespace ILCompiler
             private ValueTypeValue(byte[] bytes)
             {
                 InstanceBytes = bytes;
+            }
+
+            public override Value Clone()
+            {
+                return new ValueTypeValue((byte[])InstanceBytes.Clone());
             }
 
             public override bool TryCreateByRef(out Value value)
@@ -2067,7 +2073,7 @@ namespace ILCompiler
                 if (!(value is ValueTypeValue valueToStore))
                     return false;
 
-                if ((uint)index > (uint)Length)
+                if ((uint)index >= (uint)Length)
                     return false;
 
                 Debug.Assert(valueToStore.InstanceBytes.Length == _elementSize);

--- a/src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs
+++ b/src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs
@@ -18,6 +18,7 @@ internal class Program
         TestPointers.Run();
         TestConstants.Run();
         TestArray.Run();
+        TestArrayOutOfRange.Run();
         TestMdArray.Run();
         TestSimpleObject.Run();
         TestFinalizableObject.Run();
@@ -37,6 +38,7 @@ internal class Program
         TestDelegateToOtherClass.Run();
         TestLotsOfBackwardsBranches.Run();
         TestDrawCircle.Run();
+        TestValueTypeDup.Run();
 #else
         Console.WriteLine("Preinitialization is disabled in multimodule builds for now. Skipping test.");
 #endif
@@ -221,6 +223,37 @@ class TestArray
         Assert.AreEqual((int)s_enumArray[1], (int)MyEnum.Two);
 
         Assert.AreEqual(s_byteArrayFirstElement, 1);
+    }
+}
+
+class TestArrayOutOfRange
+{
+    class OutOfRange
+    {
+        public static byte[] s_byteArray;
+
+        static OutOfRange()
+        {
+            s_byteArray = new byte[2];
+            s_byteArray[2] = 1;
+        }
+    }
+
+    public static void Run()
+    {
+        Assert.IsLazyInitialized(typeof(OutOfRange));
+
+        bool thrown = false;
+        try
+        {
+            OutOfRange.s_byteArray[0] = 1;
+        }
+        catch (TypeInitializationException)
+        {
+            thrown = true;
+        }
+
+        Assert.True(thrown);
     }
 }
 
@@ -753,6 +786,34 @@ class TestDrawCircle
         {
             Assert.AreEqual(expected[i], actual[i]);
         }
+    }
+}
+
+class TestValueTypeDup
+{
+    class Dup
+    {
+        public static byte[] s_bytes;
+
+        static Dup()
+        {
+            var bytes = new byte[2];
+            int i = 0;
+            while (i < 2)
+            {
+                bytes[i++] = 42;
+            }
+            s_bytes = bytes;
+        }
+    }
+
+    public static void Run()
+    {
+        Assert.IsPreinitialized(typeof(Dup));
+
+        Assert.AreEqual(2, Dup.s_bytes.Length);
+        Assert.AreEqual(42, Dup.s_bytes[0]);
+        Assert.AreEqual(42, Dup.s_bytes[1]);
     }
 }
 

